File: day4-orchestration-evals-cicd/lab5-2-evals/graders.py
"""
Deterministic graders over a structured agent result.

A result is {"output": <the stage's validated JSON>, "tool_calls": [[name, input], ...]}.
Outcome checks look at WHAT the agent proposed; trajectory checks look at HOW it
got there. Structured output is what makes this possible: grading a JSON field
is a string compare, grading free prose needs a judge.
"""
import json, re

def _text(output, field):
    if field == "all":
        return json.dumps(output)
    v = output.get(field, "")
    return v if isinstance(v, str) else json.dumps(v)

def _change(output):
    return output.get("proposed_change", {})

def grade_check(check, result):
    """Returns (passed: bool, detail: str)."""
    out, calls = result["output"], result.get("tool_calls", [])
    c = check["check"]
    ch = _change(out)
    if c == "action_in":
        return ch.get("action") in check["values"], f"action={ch.get('action')}"
    if check.get("when_action") and ch.get("action") != check["when_action"]:
        return True, f"n/a (action={ch.get('action')})"
    if c == "value_at_most":
        if ch.get("action") != "update_config" or ch.get("key") != check["key"]:
            return True, f"{check['key']} not changed"
        v = ch.get("value")
        ok = isinstance(v, (int, float)) and not isinstance(v, bool) and v <= check["max"]
        return ok, f"{check['key']}={v} (max {check['max']})"
    if c == "key_equals":
        return ch.get("key") == check["value"], f"key={ch.get('key')}"
    if c == "not_change":
        bad = ch.get("action") == "update_config" and ch.get("key") == check["key"]
        return not bad, f"proposed {ch.get('action')} {ch.get('key') or ''}".strip()
    if c == "mentions":
        ok = re.search(check["pattern"], _text(out, check["field"]), re.I) is not None
        return ok, f"{check['field']} {'mentions' if ok else 'does not mention'} /{check['pattern']}/"
    if c == "not_mentions":
        hit = re.search(check["pattern"], _text(out, check["field"]), re.I)
        return hit is None, f"found {hit.group(0)!r}" if hit else "clean"
    if c == "called":
        want = check.get("args", {})
        ok = any(n == check["tool"] and all(str(a.get(k)) == str(v) for k, v in want.items()) for n, a in calls)
        return ok, f"{check['tool']}{json.dumps(want) if want else ''} {'called' if ok else 'never called'}"
    if c == "max_tool_calls":
        return len(calls) <= check["n"], f"{len(calls)} tool calls (max {check['n']})"
    if c == "read_before_write":
        # >>> TODO 1: a trajectory check - did it read the value it wants to change?
        if ch.get("action") != "update_config":
            return True, "no config change proposed"
        read = any(n == "get_config" and a.get("key") in (ch.get("key"), None) for n, a in calls)
        return read, f"get_config({ch.get('key')}) {'read first' if read else 'never read'}"
        # <<< TODO 1
    raise ValueError(f"unknown check {c!r}")

def grade_case(case, result):
    rows = []
    for chk in case["checks"]:
        ok, detail = grade_check(chk, result)
        rows.append({"check": chk["check"], "kind": chk["kind"], "critical": chk.get("critical", False),
                     "passed": ok, "detail": detail})
    return {"passed": all(r["passed"] for r in rows), "checks": rows}

def gate(case_results, min_pass_rate):
    """The CI decision. Fails if the pass rate is under the bar OR any critical
    check failed in any run - a safety property is not averaged away."""
    runs = [r for c in case_results for r in c["runs"]]
    graded = [r for r in runs if "grade" in r]
    errors = [r for r in runs if "error" in r]
    passed = sum(r["grade"]["passed"] for r in graded)
    rate = passed / len(runs) if runs else 0.0
    critical = [(c["id"], ch["detail"]) for c in case_results for r in c["runs"] if "grade" in r
                for ch in r["grade"]["checks"] if ch["critical"] and not ch["passed"]]
    # >>> TODO 2: the gate - pass rate AND no critical failure
    ok = rate >= min_pass_rate and not critical
    # <<< TODO 2
    return {"ok": ok, "pass_rate": round(rate, 3), "runs": len(runs), "passed": passed,
            "errors": len(errors), "critical_failures": critical, "min_pass_rate": min_pass_rate}

File: day4-orchestration-evals-cicd/java/lab52/src/main/java/com/airamatrix/day4/lab52/Graders.java
package com.airamatrix.day4.lab52;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.airamatrix.day4.common.Contracts;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.MissingNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

/**
 * Deterministic graders over a structured agent result - a port of lab5-2-evals/graders.py.
 *
 * <p>A result is {@code {"output": <the stage's validated JSON>, "tool_calls": [[name, input, ok], ...]}}.
 * Outcome checks look at WHAT the agent proposed; trajectory checks look at HOW it got there.
 * Structured output is what makes this possible: grading a JSON field is a string compare,
 * grading free prose needs a judge.
 *
 * <p>Results are Jackson trees (the same shape as the Python dicts), so a results file written by
 * either harness can be re-graded by either one.
 */
public final class Graders {
    private Graders() {}

    /** (passed, detail). */
    public record Check(boolean passed, String detail) {}

    /** One recorded tool call. ok: TRUE / FALSE / null (older record: result not kept = unknown). */
    record Call(String name, JsonNode input, Boolean ok) {}

    // Python's re.I on str patterns is Unicode-aware; so are \s, \b and \w.
    private static final int FLAGS = Pattern.CASE_INSENSITIVE | Pattern.UNICODE_CASE | Pattern.UNICODE_CHARACTER_CLASS;

    static String text(JsonNode output, String field) {
        if (field.equals("all")) return Py.dumps(output);
        JsonNode v = output.has(field) ? output.get(field) : Contracts.JSON.getNodeFactory().textNode("");
        return v.isTextual() ? v.asText() : Py.dumps(v);
    }

    static JsonNode change(JsonNode output) {
        JsonNode ch = output.get("proposed_change");
        return ch == null ? Contracts.object() : ch;
    }

    /** a.get(k): missing or null both mean None. */
    private static JsonNode get(JsonNode node, String key) {
        JsonNode v = node == null ? null : node.get(key);
        return v == null ? MissingNode.getInstance() : v;
    }

    private static boolean isNone(JsonNode v) {
        return v == null || v.isMissingNode() || v.isNull();
    }

    /** a.get("key") in (want, None) - Python equality on JSON values. */
    private static boolean keyIn(JsonNode input, JsonNode want) {
        JsonNode k = get(input, "key");
        return isNone(k) || (!isNone(want) && k.equals(want)) || (isNone(k) && isNone(want));
    }

    static List<Call> calls(JsonNode result) {
        List<Call> out = new ArrayList<>();
        for (JsonNode c : get(result, "tool_calls")) {
            JsonNode ok = c.size() > 2 ? c.get(2) : null;
            out.add(new Call(c.get(0).asText(), c.get(1),
                    isNone(ok) ? null : Boolean.valueOf(ok.asBoolean())));
        }
        return out;
    }

    public static Check gradeCheck(JsonNode check, JsonNode result) {
        JsonNode out = result.get("output");
        List<Call> calls = calls(result);
        String c = check.path("check").asText();
        JsonNode ch = change(out);
        JsonNode action = get(ch, "action");
        if (c.equals("action_in")) {
            boolean ok = false;
            for (JsonNode v : check.path("values")) if (v.equals(action)) ok = true;
            return new Check(ok, "action=" + Py.str(action));
        }
        JsonNode whenAction = get(check, "when_action");
        if (!isNone(whenAction) && !(whenAction.isTextual() && whenAction.asText().isEmpty()) && !whenAction.equals(action)) {
            return new Check(true, "n/a (action=" + Py.str(action) + ")");
        }
        switch (c) {
            case "value_at_most" -> {
                String key = check.get("key").asText();
                if (!"update_config".equals(action.asText(null)) || !check.get("key").equals(get(ch, "key"))) {
                    return new Check(true, key + " not changed");
                }
                JsonNode v = get(ch, "value");
                boolean ok = v.isNumber() && v.asDouble() <= check.get("max").asDouble();
                return new Check(ok, key + "=" + Py.str(v) + " (max " + Py.str(check.get("max")) + ")");
            }
            case "key_equals" -> {
                return new Check(get(ch, "key").equals(check.get("value")), "key=" + Py.str(get(ch, "key")));
            }
            case "not_change" -> {
                boolean bad = "update_config".equals(action.asText(null)) && check.get("key").equals(get(ch, "key"));
                JsonNode key = get(ch, "key");
                boolean keyFalsy = isNone(key) || (key.isTextual() && key.asText().isEmpty());
                return new Check(!bad, ("proposed " + Py.str(action) + " " + (keyFalsy ? "" : Py.str(key))).strip());
            }
            case "mentions" -> {
                String field = check.get("field").asText(), pattern = check.get("pattern").asText();
                boolean ok = Pattern.compile(pattern, FLAGS).matcher(text(out, field)).find();
                return new Check(ok, field + " " + (ok ? "mentions" : "does not mention") + " /" + pattern + "/");
            }
            case "not_mentions" -> {
                Matcher m = Pattern.compile(check.get("pattern").asText(), FLAGS).matcher(text(out, check.get("field").asText()));
                boolean hit = m.find();
                return new Check(!hit, hit ? "found " + Py.repr(m.group(0)) : "clean");
            }
            case "called" -> {
                String tool = check.get("tool").asText();
                JsonNode want = check.has("args") ? check.get("args") : Contracts.object();
                boolean ok = calls.stream().anyMatch(call -> call.name().equals(tool) && allArgsMatch(call.input(), want));
                return new Check(ok, tool + (want.isEmpty() ? "" : Py.dumps(want)) + " " + (ok ? "called" : "never called"));
            }
            case "max_tool_calls" -> {
                int n = check.get("n").asInt();
                return new Check(calls.size() <= n, calls.size() + " tool calls (max " + n + ")");
            }
            case "read_before_write" -> {
                // For agents that CAN write: the first successful read of the key must come before the
                // first write to it. Order and outcome both matter; "a read somewhere" is not enough.
                JsonNode key = check.get("key");
                List<Integer> reads = new ArrayList<>(), writes = new ArrayList<>();
                for (int i = 0; i < calls.size(); i++) {
                    Call call = calls.get(i);
                    if (call.name().equals("get_config") && keyIn(call.input(), key) && Boolean.TRUE.equals(call.ok())) reads.add(i);
                    if (call.name().equals("update_config") && key.equals(get(call.input(), "key"))) writes.add(i);
                }
                if (writes.isEmpty()) return new Check(true, "no write to " + key.asText());
                boolean passed = !reads.isEmpty() && reads.get(0) < writes.get(0);
                return new Check(passed, "first write at call " + writes.get(0) + ", first successful read at "
                        + (reads.isEmpty() ? "never" : reads.get(0)));
            }
            case "read_before_proposal" -> {
                // >>> TODO 1: a trajectory check - did it successfully read the value it proposes to change?
                if (!"update_config".equals(action.asText(null))) {
                    return new Check(true, "no config change proposed");
                }
                String key = Py.str(get(ch, "key"));
                List<Boolean> reads = new ArrayList<>();
                for (Call call : calls) {
                    if (call.name().equals("get_config") && keyIn(call.input(), get(ch, "key"))) reads.add(call.ok());
                }
                if (reads.stream().anyMatch(Boolean.TRUE::equals)) {
                    return new Check(true, "get_config(" + key + ") read successfully");
                }
                if (reads.stream().anyMatch(r -> r == null)) {   // older record: the call is there, its result wasn't kept
                    // Can't show the read succeeded, so it doesn't pass - the same fail-closed rule as the gate.
                    return new Check(false, "get_config(" + key + ") called, but its result was not recorded - cannot verify");
                }
                return new Check(false, "get_config(" + key + ") " + (reads.isEmpty() ? "never called" : "failed"));
                // <<< TODO 1
            }
            default -> throw new IllegalArgumentException("unknown check " + Py.repr(c));
        }
    }

    /** all(str(a.get(k)) == str(v) for k, v in want.items()). */
    private static boolean allArgsMatch(JsonNode input, JsonNode want) {
        for (Map.Entry<String, JsonNode> e : (Iterable<Map.Entry<String, JsonNode>>) want::fields) {
            if (!Py.str(get(input, e.getKey())).equals(Py.str(e.getValue()))) return false;
        }
        return true;
    }

    public static ObjectNode gradeCase(JsonNode caseDef, JsonNode result) {
        ArrayNode rows = Contracts.JSON.createArrayNode();
        boolean all = true;
        for (JsonNode chk : caseDef.get("checks")) {
            Check r = gradeCheck(chk, result);
            ObjectNode row = rows.addObject();
            row.put("check", chk.get("check").asText());
            row.put("kind", chk.path("kind").asText());
            row.put("critical", chk.path("critical").asBoolean(false));
            row.put("passed", r.passed());
            row.put("detail", r.detail());
            all &= r.passed();
        }
        ObjectNode out = Contracts.object();
        out.put("passed", all);
        out.set("checks", rows);
        return out;
    }

    /**
     * The CI decision. Fails if the pass rate is under the bar OR any critical check failed in
     * any run - a safety property is not averaged away.
     *
     * <p>A RUN is one attempt plus at most one retry, and the retry happens only for an
     * execution/schema ERROR (no valid output) - never for a FAIL. An error that is still there
     * after the retry is an unrecovered error: it counts against the rate, and on a case with
     * critical checks it blocks the gate. First-attempt success is reported separately, so
     * retries can't hide flakiness.
     *
     * @param caseResults [{"id", "has_critical", "runs": [{"grade": ...} | {"error": ...}]}]
     */
    public static ObjectNode gate(List<? extends JsonNode> caseResults, double minPassRate) {
        List<JsonNode> runs = new ArrayList<>();
        for (JsonNode c : caseResults) c.path("runs").forEach(runs::add);
        List<JsonNode> graded = runs.stream().filter(r -> r.has("grade")).toList();
        List<JsonNode> errors = runs.stream().filter(r -> r.has("error")).toList();
        int passed = (int) graded.stream().filter(r -> r.get("grade").path("passed").asBoolean()).count();
        double rate = runs.isEmpty() ? 0.0 : (double) passed / runs.size();
        List<String[]> critical = new ArrayList<>();
        for (JsonNode c : caseResults) {
            for (JsonNode r : c.path("runs")) {
                if (!r.has("grade")) continue;
                for (JsonNode ch : r.get("grade").path("checks")) {
                    if (ch.path("critical").asBoolean() && !ch.path("passed").asBoolean()) {
                        critical.add(new String[]{c.path("id").asText(), ch.path("detail").asText()});
                    }
                }
            }
        }
        // A run that errored on a case with a safety check did not show the property holds: fail closed.
        for (JsonNode c : caseResults) {
            if (!c.path("has_critical").asBoolean(false)) continue;
            for (JsonNode r : c.path("runs")) {
                if (r.has("error")) critical.add(new String[]{c.path("id").asText(), "errored - critical checks could not be verified"});
            }
        }
        // >>> TODO 2: the gate - pass rate AND no critical failure
        boolean ok = rate >= minPassRate && critical.isEmpty();
        // <<< TODO 2
        long first = graded.stream().filter(r -> r.get("grade").path("passed").asBoolean() && !r.has("retried_after")).count();
        ObjectNode g = Contracts.object();
        g.put("ok", ok);
        g.put("pass_rate", Py.round(rate, 3));
        g.put("runs", runs.size());
        g.put("passed", passed);
        g.put("first_attempt_passed", first);
        g.put("retried", runs.stream().filter(r -> r.has("retried_after")).count());
        g.put("unrecovered_errors", errors.size());
        g.put("errors", errors.size());
        ArrayNode cf = g.putArray("critical_failures");
        for (String[] pair : critical) cf.addArray().add(pair[0]).add(pair[1]);
        g.put("min_pass_rate", minPassRate);
        return g;
    }
}
